src/protobuf-net/ProtoReader.ReadOnlySequence.cs

```diff
@@ -333,143 +358,143 @@ private void Log(string message)
                 Debug.WriteLine("[" + LongPosition.ToString() + "] " + message);
 #endif
             }
-            [MethodImpl(MethodImplOptions.NoInlining)]
-            private int TryReadUInt32VarintWithoutMovingSlow(Read32VarintMode mode, out uint value)
-            {
-                Span<byte> span = stackalloc byte[20];
-                var available = ImplPeekBytes(span);
-                if (available != 20) span = span.Slice(0, available);
-                var read = TryParseUInt32Varint(mode == Read32VarintMode.Signed, span, 0, out value);
-                Log($"T32! - {read}:{value}");
-                if (read != 0 && mode == Read32VarintMode.FieldHeader) ReadPreviewField(value, span, read);
-                return read;
-            }
+
             private protected override int ImplTryReadUInt64VarintWithoutMoving(ref State state, out ulong value)
             {
-                var read = previewFieldBytes;
-                if (read != 0)
+                if (previewFieldBytes != 0)
                 {
+                    var read = previewFieldBytes;
                     value = checked((uint)PeekPreviewField());
                     Log($"T64* - {read}:{value}");
                     return read;
                 }
 
                 if (GetSomeData(false) >= 10)
                 {
-                    read = TryParseUInt64Varint(Peek(out var offset), offset, out value);
+                    var read = TryParseUInt64Varint(Peek(out var offset), offset, out value);
                     Log($"T64 - {read}:{value}");
                     return read;
                 }
+                else
+                {
+                    return ViaStackAlloc(out value);
+                }
 
-                Span<byte> span = stackalloc byte[10];
-                read = ImplPeekBytes(span);
-                if (read != 10) span = span.Slice(0, read);
-                read = TryParseUInt64Varint(span, 0, out value);
-                Log($"T64! - {read}:{value}");
-                return read;
+                int ViaStackAlloc(out ulong val)
+                {
+                    Span<byte> span = stackalloc byte[10];
+                    var read = ImplPeekBytes(span);
+                    if (read != 10) span = span.Slice(0, read);
+                    read = TryParseUInt64Varint(span, 0, out val);
+                    Log($"T64! - {read}:{val}");
+                    return read;
+                }
             }
+
+
             private int TryParseUInt32Varint(bool trimNegative, ReadOnlySpan<byte> span, int offset, out uint value)
             {
-                var available = span.Length - offset;
-                if (available == 0)
+                if (0 >= (uint)span.Length)
                 {
                     value = 0;
                     return 0;
                 }
-                value = span[offset++];
+
+                value = span[0];
```

Comment:
I'm concerned that offset has been lost here; the intent of offset was to avoid slicing in common in-situ scenarios... it looks like this will read the wrong data?

Reply:
Ah, copy paste issue :)